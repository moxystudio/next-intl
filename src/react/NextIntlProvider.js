import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { RawIntlProvider, createIntl, createIntlCache } from 'react-intl';
import NextIntlContext from './util/context';
import createManager, { getInitialData } from '../manager';

const Provider = NextIntlContext.Provider;

export default class NextIntlProvider extends PureComponent {
    static propTypes = {
        locales: PropTypes.arrayOf(PropTypes.shape({
            name: PropTypes.string.isRequired,
            id: PropTypes.string.isRequired,
            loadMessages: PropTypes.func.isRequired,
        })),
        policies: PropTypes.arrayOf(PropTypes.shape({
            match: PropTypes.func.isRequired,
            watch: PropTypes.func,
            save: PropTypes.func,
            act: PropTypes.func,
        })),
        initialData: PropTypes.shape({
            localeId: PropTypes.string.isRequired,
            messages: PropTypes.object.isRequired,
        }),
        onChange: PropTypes.func,
        children: PropTypes.node,
    };

    state;
    manager;
    intlCache = createIntlCache();

    constructor(props) {
        super(props);

        this.buildManager();
        this.state = this.buildState();
    }

    componentDidUpdate(prevProps, prevState) {
        if (prevProps.locales !== this.props.locales ||
            prevProps.policies !== this.props.policies) {
            this.buildManager();
            this.setState(this.buildState());
        }

        if (prevState !== this.state) {
            this.props.onChange?.(this.state);
        }
    }

    componentWillUnmount() {
        this.manager.destroy();
    }

    render() {
        const { children } = this.props;
        const value = this.state;

        return (
            <Provider value={ value }>
                <RawIntlProvider value={ value.intl }>
                    { children }
                </RawIntlProvider>
            </Provider>
        );
    }

    getValue() {
        return this.state;
    }

    buildManager() {
        const { locales, policies, initialData } = this.props;
        const data = this.manager ? this.manager.toData() : initialData;

        this.manager?.destroy();

        this.manager = createManager(locales, policies, data);
        this.manager.onLocaleChange(() => this.setState(this.buildState()));
    }

    buildState() {
        const { locales, policies, initialData, onChange, children, ...intlConfig } = this.props;

        const { locale, messages, changeLocale } = this.manager;
        const intl = createIntl({
            ...intlConfig,
            locale: locale.id,
            messages,
        });

        return {
            locales,
            locale,
            changeLocale,
            intl,
        };
    }
}

export const getInitialProps = async (config, ctx) => {
    const { locales, policies } = config;

    const initialData = await getInitialData(locales, policies, ctx);

    return { initialData };
};
