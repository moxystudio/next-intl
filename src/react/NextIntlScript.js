import fs from 'fs';
import React from 'react';
import memoizeOne from 'memoize-one';
import PropTypes from 'prop-types';

// The script bellow was minified using terser: https://xem.github.io/terser-online/
// You may find the original script below in case you need to modify it:
/*
(function () {
if (!Intl.PluralRules || !Intl.RelativeTimeFormat) {
    var currentScript = document.currentScript;
    var polyfillScript = document.createElement('script');

    polyfillScript.src = '${polyfillChunkPublicPath}';

    currentScript.parentNode.insertBefore(polyfillScript, currentScript.nextSibling);
})();
*/
const buildScript = (polyfillChunkPublicPath) => `(function(){if(!Intl.PluralRules||!Intl.RelativeTimeFormat){var e=document.currentScript,t=document.createElement("script");t.src="${polyfillChunkPublicPath}",e.parentNode.insertBefore(t,e.nextSibling)}})();`;

const getPolyfillChunkPublicPath = (assetPrefix) => {
    const manifest = JSON.parse(fs.readFileSync('.next/react-loadable-manifest.json'));
    const publicPath = manifest?.['@formatjs/intl-pluralrules/polyfill-locales']?.[0]?.publicPath;

    if (!publicPath) {
        throw new Error('Could not find intl-polyfill chunk in .next/react-loadable-manifest.json');
    }

    return `${assetPrefix}/_next/${publicPath}`;
};

const memoizedGetPolyfillChunkPublicPath =
    process.env.NODE_ENV === 'test' ? getPolyfillChunkPublicPath : memoizeOne(getPolyfillChunkPublicPath);

const NextIntlScript = ({ assetPrefix }) => {
    const polyfillChunkPublicPath = memoizedGetPolyfillChunkPublicPath(assetPrefix);
    const script = buildScript(polyfillChunkPublicPath);

    return <script dangerouslySetInnerHTML={ { __html: script } } />;
};

NextIntlScript.defaultProps = {
    assetPrefix: '',
};

NextIntlScript.propTypes = {
    assetPrefix: PropTypes.string,
};

export default NextIntlScript;
