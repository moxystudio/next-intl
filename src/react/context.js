import React, { createContext, PureComponent } from 'react';
import PropTypes from 'prop-types';
import { IntlProvider } from 'react-intl';
import createManager from '../manager';

const NextIntlContext = createContext();
const Provider = NextIntlContext.Provider;

class NextIntlProvider extends PureComponent {
    state = {};
    manager;

    constructor(props) {
        super(props);

        this.buildManager();
        this.state = this.buildState();
    }

    componentDidUpdate(prevProps) {
        if (prevProps.locales !== this.props.locales ||
            prevProps.policies !== this.props.policies) {
            this.buildManager();
            this.setState(this.buildState());
        }
    }

    render() {
        const { locales, children, ...rest } = this.props;
        const { providerValue, intlProps } = this.state;

        return (
            <Provider value={ providerValue }>
                <IntlProvider { ...rest } { ... intlProps }>
                    { children }
                </IntlProvider>
            </Provider>
        );
    }

    buildManager() {
        const { locales, policies, initialData } = this.props;
        const data = this.manager ? this.manager.toData() : initialData;

        this.manager.destroy();

        this.manager = createManager(locales, policies, data);
        this.manager.onLocaleChange(() => this.setState(this.buildState()));
    }

    buildState() {
        const { locales } = this.props;

        return {
            providerValue: {
                locales,
                locale: this.manager.locale,
                changeLocale: this.manager.changeLocale,
            },
            intlProps: {
                locale: this.manager.locale.id,
                messages: this.manager.messages,
            },
        };
    }
}

NextIntlProvider.propTypes = {
    locales: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string.isRequired,
        id: PropTypes.string.isRequired,
        loadMessages: PropTypes.func.isRequired,
    })),
    policies: PropTypes.arrayOf(PropTypes.shape({
        match: PropTypes.func.isRequired,
        watch: PropTypes.func,
        save: PropTypes.func,
        act: PropTypes.func,
    })),
    initialData: PropTypes.shape({
        localeId: PropTypes.string.isRequired,
        messages: PropTypes.object.isRequired,
    }).isRequired,
    children: PropTypes.node,
};

NextIntlContext.Provider = NextIntlProvider;
NextIntlContext.Consumer.Provider = NextIntlProvider;

export default NextIntlContext;
