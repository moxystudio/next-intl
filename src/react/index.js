import NextIntlContext from './context';

const { Provider: NextIntlProvider, Consumer: NextIntlConsumer } = NextIntlContext;

export { NextIntlContext, NextIntlProvider, NextIntlConsumer };
export { default as useNextIntl } from './hook';
export { default as NextIntlScript } from './script';
