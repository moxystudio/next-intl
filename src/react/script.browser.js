const NextIntlScript = () => null;

export default NextIntlScript;
