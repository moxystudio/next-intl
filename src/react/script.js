import fs from 'fs';
import React from 'react';

// The script bellow was minified using terser: https://xem.github.io/terser-online/
// You may find the original script below in case you need to modify it:
/*
(function () {
if (!Intl.PluralRules || !Intl.RelativeTimeFormat) {
    var currentScript = document.currentScript;
    var polyfillScript = document.createElement('script');

    polyfillScript.src = '${polyfillChunkPublicPath}';

    currentScript.parentNode.insertBefore(polyfillScript, currentScript.nextSibling);
})();
*/
const buildScript = (polyfillChunkPublicPath) => `(function(){if(!Intl.PluralRules||!Intl.RelativeTimeFormat){var e=document.currentScript,t=document.createElement("script");t.src="${polyfillChunkPublicPath}",e.parentNode.insertBefore(t,e.nextSibling)}})();`;

const getPolyfillChunkPublicPath = () => {
    const manifest = JSON.parse(fs.readFileSync('.next/react-loadable-manifest.json'));
    const publicPath = manifest?.['@formatjs/intl-pluralrules/polyfill-locales']?.[0]?.publicPath;

    if (!publicPath) {
        throw new Error('Could not find intl-polyfill chunk in .next/react-loadable-manifest.json');
    }

    return __webpack_public_path__ + publicPath; // eslint-disable-line
};

const NextIntlScript = () => {
    const polyfillChunkPublicPath = getPolyfillChunkPublicPath();

    const script = buildScript(polyfillChunkPublicPath);

    return <script dangerouslySetInnerHTML={ { __html: script } } />;
};

export default NextIntlScript;
