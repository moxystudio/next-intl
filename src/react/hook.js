import { useContext } from 'react';
import NextIntlContext from './context';

const useNextIntl = () => useContext(NextIntlContext);

export default useNextIntl;
