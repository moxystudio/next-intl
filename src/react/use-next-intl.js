import { useContext } from 'react';
import NextIntlContext from './util/context';

const useNextIntl = () => useContext(NextIntlContext);

export default useNextIntl;
