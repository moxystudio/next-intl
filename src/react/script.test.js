import fs from 'fs';
import React from 'react';
import { render } from '@testing-library/react';
import NextIntlScript from './script';

jest.mock('fs', () => ({
    readFileSync: jest.fn(),
}));

beforeAll(() => {
    global.__webpack_public_path__ = '/_next/'; // eslint-disable-line
});

it('should render a script tag with conditional loading of the polyfill correctly', () => {
    fs.readFileSync.mockImplementation(() => (
        JSON.stringify({
            '@formatjs/intl-pluralrules/polyfill-locales': [
                { publicPath: 'static/chunks/intl-polyfill.js' },
            ],
        })),
    );

    const { container } = render(
        <NextIntlScript />,
    );

    const script = container.querySelector('script');

    expect(script.innerHTML).toMatchSnapshot();
});

it('should fail if unable to find polyfill chunk', () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});

    fs.readFileSync.mockImplementation(() => '{}');

    expect(() => {
        render(
            <NextIntlScript />,
        );
    }).toThrow(new Error('Could not find intl-polyfill chunk in .next/react-loadable-manifest.json'));
});
