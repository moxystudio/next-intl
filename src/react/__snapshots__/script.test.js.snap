// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a script tag with conditional loading of the polyfill correctly 1`] = `"(function(){if(!Intl.PluralRules||!Intl.RelativeTimeFormat){var e=document.currentScript,t=document.createElement(\\"script\\");t.src=\\"/_next/static/chunks/intl-polyfill.js\\",e.parentNode.insertBefore(t,e.nextSibling)}})();"`;
